# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/GCParallelTask.cpp
# Commit: 2fa53cd9ac4e
# Full Hash: 2fa53cd9ac4e821763dbc4402e58b58184ce230b
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-03-04 04:18:28
# Regressor Bug: 1643586
# File Overlap Count: 1
# Description:
#   Bug 1643586 - Part 6: Stop using JSContext for GC helper thread tasks r=jandem
#   
#   This patch moves the debug-only GC state from JSContext to JSFreeOp. Support
#   functions for this move to FreeOp.h and most RAII classes move to GCInternals.h.
#   
# ==============================================================================

diff -r 054a08d13999 -r 2fa53cd9ac4e js/src/gc/GCParallelTask.cpp
--- a/js/src/gc/GCParallelTask.cpp	Thu Mar 03 17:49:18 2022 +0000
+++ b/js/src/gc/GCParallelTask.cpp	Thu Mar 03 17:49:18 2022 +0000
@@ -9,6 +9,8 @@
 #include "mozilla/MathAlgorithms.h"
 #include "mozilla/Maybe.h"
 
+#include "gc/FreeOp.h"
+#include "gc/GCInternals.h"
 #include "gc/ParallelWork.h"
 #include "vm/HelperThreadState.h"
 #include "vm/Runtime.h"
@@ -156,12 +158,15 @@
 
   setRunning(lock);
 
-  JS::ContextOptions options;  // The GC doesn't use any context options.
-  AutoSetHelperThreadContext usesContext(options, lock);
-  AutoSetContextRuntime ascr(gc->rt);
+  JSFreeOp freeOp(gc->rt, false);
+  MOZ_RELEASE_ASSERT(TlsFreeOp.init());
+  TlsFreeOp.set(&freeOp);
+
   gc::AutoSetThreadIsPerformingGC performingGC;
   runTask(lock);
 
+  TlsFreeOp.set(nullptr);
+
   setFinished(lock);
 }
 